lib/hdt.js

```diff
@@ -54,19 +54,38 @@ HdtDocumentPrototype.searchTerms = function (options) {
   if (this.closed) return Promise.reject(new Error('The HDT document cannot be read because it is closed'));
   const limit = options && options.limit ? Math.max(0, parseInt(options.limit, 10)) : 100,
       position = options && options.position,
+      posId = POSITIONS[position],
       prefix = options && options.prefix || '',
-      posId = POSITIONS[position];
+      filterPos = options.filter && options.filter.position,
+      filterVal = options.filter && options.filter.value,
+      filterPosId = POSITIONS[filterPos];
   if (!(position in POSITIONS))
     return Promise.reject(new Error('Invalid position argument. Expected subject, predicate or object.'));
+
   return new Promise((resolve, reject) => {
-    this._searchTerms(prefix, limit, posId, (error, results) => {
-      if (error)
-        reject(error);
-      else
-        resolve(results);
-    }, this);
+    if (options.filter) {
+      if (!filterVal || filterVal === '')
+       return reject(new Error('Unspecified filter value.'));
+      if (filterPos !== 'object' || position !== 'predicate')
+         return reject(new Error('Unsupported filter position argument.'));
+      this._fetchDistinctTerms(filterVal, filterPosId, limit, posId, (error, results) => {
+        if (error)
+          reject(error);
+        else
+          resolve(results);
+      }, this);
```

Comment:
Inline the above as `(error, results) => error ? reject(error) : resolve(results)` (small ideas should have small code).